src/number_theory/number_field/canonical_embedding.lean

```diff
@@ -0,0 +1,191 @@
+/-
+Copyright (c) 2022 Xavier Roblot. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Xavier Roblot
+-/
+
+import number_theory.number_field.embeddings
+
+/-!
+# Canonical embedding of a number field
+The canonical embedding of a number field `K` of signature `(r₁, r₂)` is the ring homomorphism
+`K →+* ℝ^r₁ × ℂ^r₂` that sends `x ∈ K` to `(φ_₁(x),...,φ_r₁(x)) × (ψ_₁(x),..., ψ_r₂(x))` where
+`φ_₁,...,φ_r₁` are its real embeddings and `ψ_₁,..., ψ_r₂` are its complex embeddings (up to
+complex conjugation).
+
+## Main definitions and results
+* `number_field.canonical_embedding.ring_of_integers.inter_ball_finite`: the intersection of the
+image of the ring of integers by the canonical embedding and any ball centered at `0` of finite
+radius is finite.
+
+## Tags
+number field, infinite places
+-/
+
+open_locale classical number_field
+
+noncomputable theory
+
+open number_field number_field.infinite_place module fintype finite_dimensional
+
+variables (K : Type*) [field K]
+
+namespace number_field.canonical_embedding
+
+/-- The ambiant space `ℝ^r₁ × ℂ^r₂` with `(r₁, r₂)` the signature of `K`. -/
+@[reducible]
+def space :=
+  ({w : infinite_place K // is_real w} → ℝ) × ({w : infinite_place K // is_complex w} → ℂ)
+
+instance : comm_ring (space K) := prod.comm_ring
+
+instance : module ℝ (space K) := prod.module
+
+lemma space_rank [number_field K] :
+  finrank ℝ (space K) = finrank ℚ K :=
+begin
+  haveI : module.free ℝ ℂ := infer_instance,
+  rw [module.free.finrank_prod, module.free.finrank_pi, module.free.finrank_pi_fintype,
+    complex.finrank_real_complex, finset.sum_const, finset.card_univ, ← card_real_embeddings,
+    algebra.id.smul_eq_mul, mul_comm, ← card_complex_embeddings, ← number_field.embeddings.card K ℂ,
+    fintype.card_subtype_compl, nat.add_sub_of_le (fintype.card_subtype_le _)],
+end
+
+lemma space_nontrivial [number_field K] : nontrivial (space K) :=
```

Comment:
Can't this be an instance?

Reply:
I changed this to an `instance` but it triggers the ``fails_quickly` linter. After investigating, it just that Lean takes a very long time to decide that instance fails. So I added `@[nolint fails_quickly]` to the instance.